homeassistant/components/harmony/remote.py

```diff
@@ -405,94 +319,21 @@ async def async_send_command(self, command, **kwargs):
-            "Sending commands to device %s holding for %s seconds "
-            "with a delay of %s seconds",
-            device,
-            hold_secs,
-            delay_secs,
+        await self._data.async_send_command(
+            command, device, num_repeats, delay_secs, hold_secs
         )
 
-        # Creating list of commands to send.
-        snd_cmnd_list = []
-        for _ in range(num_repeats):
-            for single_command in command:
-                send_command = SendCommandDevice(
-                    device=device_id, command=single_command, delay=hold_secs
-                )
-                snd_cmnd_list.append(send_command)
-                if delay_secs > 0:
-                    snd_cmnd_list.append(float(delay_secs))
-
-        _LOGGER.debug("%s: Sending commands", self.name)
-        try:
-            result_list = await self._client.send_commands(snd_cmnd_list)
-        except aioexc.TimeOut:
-            _LOGGER.error("%s: Sending commands timed-out", self.name)
-            return
-
-        for result in result_list:
-            _LOGGER.error(
-                "Sending command %s to device %s failed with code %s: %s",
-                result.command.command,
-                result.command.device,
-                result.code,
-                result.msg,
-            )
-
     async def change_channel(self, channel):
         """Change the channel using Harmony remote."""
-        _LOGGER.debug("%s: Changing channel to %s", self.name, channel)
-        try:
-            await self._client.change_channel(channel)
-        except aioexc.TimeOut:
-            _LOGGER.error("%s: Changing channel to %s timed-out", self.name, channel)
+        self._data.change_channel(channel)
 
     async def sync(self):
         """Sync the Harmony device with the web service."""
-        _LOGGER.debug("%s: Syncing hub with Harmony cloud", self.name)
-        try:
-            await self._client.sync()
-        except aioexc.TimeOut:
-            _LOGGER.error("%s: Syncing hub with Harmony cloud timed-out", self.name)
-        else:
-            await self.hass.async_add_executor_job(self.write_config_file)
-
-    def write_config_file(self):
-        """Write Harmony configuration file."""
-        _LOGGER.debug(
-            "%s: Writing hub configuration to file: %s", self.name, self._config_path
-        )
-        if self._client.config is None:
-            _LOGGER.warning("%s: No configuration received from hub", self.name)
-            return
+        await self._data.sync()
 
-        try:
-            with open(self._config_path, "w+", encoding="utf-8") as file_out:
-                json.dump(self._client.json_config, file_out, sort_keys=True, indent=4)
-        except OSError as exc:
-            _LOGGER.error(
-                "%s: Unable to write HUB configuration to %s: %s",
-                self.name,
-                self._config_path,
-                exc,
-            )
+    async def sleep(self, time):
```

Comment:
Had better luck splitting this into a specific method I could patch rather than patching `asyncio.sleep` directly in the connection state change tests. I was probably missing something subtle in how I was attempting to mock it away, so apologies if this is not a best practice.